### PR TITLE
TypedArray#set

### PR DESCRIPTION
**Pull request by jtenner**
This implementation tries to take heavy advantage of conditional compilation to enable array parameters.

# Accomplishes

- Setting from array of values
- Setting from TypedArray
- Setting from ArrayLike classes
- Conversion of data types
- Bypass tslint problems

The test suite passes, and I'm open to optimization. I know that speed is important, but because `Array.isView()` does not compile to a constant, it's hard to take advantage of conditional compilation in this case.

# Issues: 

Currently the only way to determine if the underlying access type of `SourceT` when it is an array, we must actually perform a (un)checked access to grab a concrete variable with the correct type `SourceU`. Without ever referencing it, we must rely on `stub instanceof [numbertype]` to validate what the source is, then use the `SET_COPY()` inline method to actually move the data.

Please feel free to help me fix this right away, because there is a very large demand to get this function working. 👍

### CODE REVIEW

File: std/assembly/internal/typedarray.ts [hunk 1]
```diff
@@ -231,3 +231,135 @@ export function EVERY<TArray extends TypedArray<T>, T>(
...
+@inline
+export function SET<T extends TypedArray<U>, U extends number, SourceT, SourceU extends number>(
+  target: T,
+  source: SourceT,
+  offset: i32): void {
+  // assert target and source are not null
+  if (isReference<T>(target)) {
+    assert(target != null, "TypeError: Target is null.");
+  } else {
+    assert(false, "TypeError: T is not a reference.");
```
Comment (MaxGraey) [inline std/assembly/internal/typedarray.ts]: `"TypeError: target is not a reference."`

File: std/assembly/internal/typedarray.ts [hunk 2]
```diff
@@ -231,3 +231,135 @@ export function EVERY<TArray extends TypedArray<T>, T>(
...
+  }
+
+  if (isReference<SourceT>(source)) {
+    assert(source != null, "TypeError: Source is null.");
+  } else {
+    assert(false, "TypeError: SourceT is not a reference.");
```
Comment (MaxGraey) [inline std/assembly/internal/typedarray.ts]: "TypeError: source is not a reference."

**Comment by jtenner** [comment]
There are a bunch of instances where the typescript compiler does not like the use of `any`. There are currently no workarounds for this.

File: std/assembly/internal/typedarray.ts [hunk 3]
```diff
@@ -231,3 +231,135 @@ export function EVERY<TArray extends TypedArray<T>, T>(
...
+      // cache the buffer and the byteOffset
+      let targetBuffer = target.buffer;
+      let targetByteOffset = target.byteOffset;
+
+      // for each source value, write it to the ArrayBuffer
+      for (let i = 0; i < sourceLength; i++) {
+        STORE<U>(
+          targetBuffer,
+          i + offset,
+          <U>unchecked(source[i]),
```
Comment (MaxGraey) [inline std/assembly/internal/typedarray.ts]: use `LOAD` over `source.buffer_`

Comment (jtenner) [reply std/assembly/internal/typedarray.ts]: Oh that's right! I think I can normalize a lot of this.

File: std/assembly/typedarray.ts [hunk 4]
```diff
@@ -109,6 +114,10 @@ export class Uint8Array extends TypedArray<u8> {
   every(callbackfn: (value: u8, index: i32, self: Uint8Array) => bool): bool {
     return EVERY<Uint8Array, u8>(this, callbackfn);
   }
+
+  set<SourceT, SourceU>(source: SourceT, offset: i32 = 0): void {
```
Comment (MaxGraey) [inline std/assembly/typedarray.ts]: Also using `<SourceT, SourceU>` looks redundant. It should be `set<T>(source: T, offset: i32 = 0)`

Comment (jtenner) [reply std/assembly/typedarray.ts]: We cannot infer what the subtype is. It looks redundant, and is not fixable. Let me commit again and show you what I mean.

File: std/assembly/internal/typedarray.ts (hunk 3 above)
Comment (jtenner) [reply std/assembly/internal/typedarray.ts]: This should be resolved with the latest commit.

File: std/assembly/typedarray.ts (hunk 4 above)
Comment (jtenner) [reply std/assembly/typedarray.ts]: If we want to use the `LOAD<T>` macro for arraybuffers, we must concede that we need a `SourceU` type.

There are many ways to do this in TypeScript, but AssemblyScript does not support this yet. Instead we need to declare the type up front. This is redundant, and feels wrong, but it's the only way to do this.

Comment (jtenner) [reply std/assembly/typedarray.ts]: I used a stub and instanceof. I don't know if this works well.

**Comment by dcodeIO** [comment]
Looks like we'd need

1. `typeof EXPR` evaluating to the type of any expression, which in turn requires that the resolver is able to resolve any expression without first compiling it [as mentioned here](https://github.com/AssemblyScript/assemblyscript/issues/473) and
2. type aliases within function bodies (that's currently limited to top-level) to make this easier to work with

**Comment by jtenner** [comment]
Yeah. I also need to add a slowpath check and see if the array is writing to itself. I need a memory allocation.

**Comment by jtenner** [comment]
This is now ready for review again.

# Current problems

1. No static way to determine if class is ArrayLike. Wishlist:
  - Need a way to get `typeof` without instanceof stub value
  - Could use static compiler constant `isArrayLike()` (and some way to check and see if it uses `unchecked()` gets?)
2. Code looks difficult to maintain because of many instanceof checks (which would be solved by solving 1)
3. Too many useless comments, and probably not good enough comments on parts that need it

**Comment by jtenner** [comment]
I vote that we close this pull request and wait for better type support.

**Comment by vgrichina** [comment]
@dcodeIO @MaxGraey is it hard to add necessary features to compiler?

**Comment by dcodeIO** [comment]
Adding `typeof` is somewhat blocked by the resolver not yet being able to resolve any kind of expression without first compiling it (second part of #473), which is not necessarily hard but quite some work. Certainly will get to it eventually.

**Comment by dcodeIO** [comment]
Have been thinking about this a bit lately, and I figured that maybe RTTI can be used to implement it. Target properties are known statically

```ts
const targetAlignLog2 = alignof<T>();
const targetIsFloat = isFloat<T>();
```

while source properties can be obtained from RTTI:

```ts
var sourceInfo = __typeinfo(changetype<BLOCK>(source).rtId);
var sourceAlignLog2: usize = 31 - Math.clz32((sourceInfo / TypeinfoFlags.VALUE_ALIGN_0) & 31);
var sourceIsFloat = (sourceInfo & TypeinfoFlags.VALUE_FLOAT) != 0;
var sourceIsManaged = (sourceInfo & TypeinfoFlags.VALUE_MANAGED) != 0; // disallow this
```

That'd then give us all the possible combinations of integers and floating point values, even for normal arrays. Challenge is to statically eliminate as much as possible by checking target first.

**Comment by jtenner** [comment]
Does that solution account for Uint8ClampedArray?

**Comment by dcodeIO** [comment]
That's a special case, yeah, but it's only relevant when writing values, which is possible to check statically.

**Comment by jtenner** [comment]
Alright. The way I see it, I can try to implement this again, or someone can take up the responsibility. It's father's day this weekend and I'll be out camping.

Since I don't know enough about rtti, it may be best for @MaxGraey or yourself to implement this feature.

**Pull request closed without merging**